File: Chap6-GraphicsViewFramework/6-12FlappyBirdQt/birditem.cpp
#include "birditem.h"
#include <QTimer>
#include <QGraphicsScene>

BirdItem::BirdItem(QPixmap pixmap)  :
    wingPosition(WingPosition::Up),
    wingDirection(0)
{
    setPixmap(pixmap);

    QTimer * birdWingsTimer = new QTimer(this);
    connect(birdWingsTimer,&QTimer::timeout,[=](){
        updatePixmap();
    });

    birdWingsTimer->start(80);

    groundPosition = scenePos().y() + 290;

    yAnimation = new QPropertyAnimation(this,"y",this);
    yAnimation->setStartValue(scenePos().y());
    yAnimation->setEndValue(groundPosition);
    yAnimation->setEasingCurve(QEasingCurve::InQuad);
    yAnimation->setDuration(1000);

  //  yAnimation->start();

    rotationAnimation = new QPropertyAnimation(this,"rotation",this);
   // rotateTo(90,1200,QEasingCurve::InQuad);

}

qreal BirdItem::rotation() const
{
    return m_rotation;
}

qreal BirdItem::y() const
{
    return m_y;
}

void BirdItem::shootUp()
{
    yAnimation->stop();
    rotationAnimation->stop();

    qreal curPosY = y();

    yAnimation->setStartValue(curPosY);
    yAnimation->setEndValue(curPosY - scene()->sceneRect().height()/8);
    yAnimation->setEasingCurve(QEasingCurve::OutQuad);
    yAnimation->setDuration(285);

    connect(yAnimation,&QPropertyAnimation::finished,[=](){
       fallToGroundIfNecessary();
    });


    yAnimation->start();

    rotateTo(-20,200,QEasingCurve::OutCubic);
}

void BirdItem::startFlying()
{
    yAnimation->start();
    rotateTo(90,1200,QEasingCurve::InQuad);
}

void BirdItem::freezeInPlace()
{
    yAnimation->stop();
    rotationAnimation->stop();
}

void BirdItem::setRotation(qreal rotation)
{
    m_rotation = rotation;

    QPointF c = boundingRect().center();

    QTransform t;
    t.translate(c.x(), c.y());
    t.rotate(rotation);
    t.translate(-c.x(), -c.y());
    setTransform(t);
}

void BirdItem::setY(qreal y)
{
    moveBy(0,y-m_y);
    m_y = y;

}

void BirdItem::rotateTo(const qreal &end, const int &duration, const QEasingCurve &curve)
{
    rotationAnimation->setStartValue(rotation());
    rotationAnimation->setEndValue(end);
    rotationAnimation->setEasingCurve(curve);
    rotationAnimation->setDuration(duration);

    rotationAnimation->start();
}

void BirdItem::fallToGroundIfNecessary()
{
    if( y() < groundPosition){

        rotationAnimation->stop();
        //yAnimation->stop();

        yAnimation->setStartValue(y());
        yAnimation->setEasingCurve(QEasingCurve::InQuad);
        yAnimation->setEndValue(groundPosition);
        yAnimation->setDuration(1200);
        yAnimation->start();

        rotateTo(90,1100,QEasingCurve::InCubic);
    }

}

void BirdItem::updatePixmap()
{
    if(wingPosition == WingPosition::Middle){

        if(wingDirection){
            //Up
            setPixmap(QPixmap(":/images/bird_blue_up.png"));
            wingPosition = WingPosition::Up;
            wingDirection = 0;
        }else{
            //Down
            //Up
            setPixmap(QPixmap(":/images/bird_blue_down.png"));
            wingPosition = WingPosition::Down;
            wingDirection = 1;
        }

    }else{
        setPixmap(QPixmap(":/images/bird_blue_middle.png"));
        wingPosition = WingPosition::Middle;
    }

}

File: Chap6-GraphicsViewFramework/6-12FlappyBirdQt/birditem.py
from PySide6.QtCore import QObject, Property, QPropertyAnimation, QEasingCurve
from PySide6.QtGui import QPixmap, QTransform
from PySide6.QtWidgets import QGraphicsPixmapItem

class BirdItem(QObject, QGraphicsPixmapItem):
    # Enum for wing positions
    class WingPosition:
        Up = 0
        Middle = 1
        Down = 2
    
    def __init__(self, pixmap):
        QObject.__init__(self)
        QGraphicsPixmapItem.__init__(self, pixmap)
        
        self.wing_position = self.WingPosition.Up
        self.wing_direction = 0  # 0: down, 1: up
        self.m_rotation = 0
        self.m_y = 0
        
        # Set up wing flapping animation timer
        from PySide6.QtCore import QTimer
        bird_wings_timer = QTimer(self)
        bird_wings_timer.timeout.connect(self.update_pixmap)
        bird_wings_timer.start(80)
        
        # Store ground position
        self.ground_position = self.scenePos().y() + 290
        
        # Y position animation
        self.y_animation = QPropertyAnimation(self, b"y", self)
        self.y_animation.setStartValue(self.scenePos().y())
        self.y_animation.setEndValue(self.ground_position)
        self.y_animation.setEasingCurve(QEasingCurve.InQuad)
        self.y_animation.setDuration(1000)
        
        # Rotation animation
        self.rotation_animation = QPropertyAnimation(self, b"rotation", self)
    
    def get_rotation(self):
        return self.m_rotation
    
    def get_y(self):
        return self.m_y
    
    def shoot_up(self):
        self.y_animation.stop()
        self.rotation_animation.stop()
        
        cur_pos_y = self.get_y()
        
        self.y_animation.setStartValue(cur_pos_y)
        # Make the bird jump higher (1/6 of screen height instead of 1/8)
        self.y_animation.setEndValue(cur_pos_y - self.scene().sceneRect().height()/6)
        self.y_animation.setEasingCurve(QEasingCurve.OutQuad)
        # Slightly longer jump duration for smoother movement
        self.y_animation.setDuration(300)
        
        self.y_animation.finished.connect(self.fall_to_ground_if_necessary)
        
        self.y_animation.start()
        
        self.rotate_to(-20, 200, QEasingCurve.OutCubic)
    
    def start_flying(self):
        self.y_animation.start()
        self.rotate_to(90, 1200, QEasingCurve.InQuad)
    
    def freeze_in_place(self):
        self.y_animation.stop()
        self.rotation_animation.stop()
    
    def set_rotation(self, rotation):
        self.m_rotation = rotation
        
        c = self.boundingRect().center()
        
        t = QTransform()
        t.translate(c.x(), c.y())
        t.rotate(rotation)
        t.translate(-c.x(), -c.y())
        self.setTransform(t)
    
    def set_y(self, y):
        self.moveBy(0, y - self.m_y)
        self.m_y = y
    
    def rotate_to(self, end, duration, curve):
        self.rotation_animation.setStartValue(self.get_rotation())
        self.rotation_animation.setEndValue(end)
        self.rotation_animation.setEasingCurve(curve)
        self.rotation_animation.setDuration(duration)
        
        self.rotation_animation.start()
    
    def fall_to_ground_if_necessary(self):
        if self.get_y() < self.ground_position:
            self.rotation_animation.stop()
            
            self.y_animation.setStartValue(self.get_y())
            self.y_animation.setEasingCurve(QEasingCurve.InQuad)
            self.y_animation.setEndValue(self.ground_position)
            # Slower falling for more reaction time
            self.y_animation.setDuration(1600)
            self.y_animation.start()
            
            self.rotate_to(90, 1100, QEasingCurve.InCubic)
    
    def update_pixmap(self):
        if self.wing_position == self.WingPosition.Middle:
            if self.wing_direction:
                # Up
                self.setPixmap(QPixmap(":/images/bird_blue_up.png"))
                self.wing_position = self.WingPosition.Up
                self.wing_direction = 0
            else:
                # Down
                self.setPixmap(QPixmap(":/images/bird_blue_down.png"))
                self.wing_position = self.WingPosition.Down
                self.wing_direction = 1
        else:
            self.setPixmap(QPixmap(":/images/bird_blue_middle.png"))
            self.wing_position = self.WingPosition.Middle
    
    # Define properties for animation
    rotation = Property(float, get_rotation, set_rotation)
    y = Property(float, get_y, set_y)